File: zenscript-language-server/src/main/java/raylras/zen/lsp/bracket/RpcClient.java
package raylras.zen.lsp.bracket;

import org.eclipse.lsp4j.jsonrpc.Launcher;
import org.eclipse.lsp4j.jsonrpc.services.JsonRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import raylras.zen.lsp.StandardIOLauncher;

import java.io.IOException;
import java.net.Socket;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.function.Function;
import java.util.stream.Collectors;

public class RpcClient {

    private static final Logger logger = LoggerFactory.getLogger(RpcClient.class);

    static Map<String, List<String>> getEntryPropertiesRemote(String validExpr)
            throws IOException, ExecutionException, InterruptedException {
        Map<String, Object> properties = getRemoteService().query(validExpr, true)
                .exceptionally(e -> {
                    invalidateRemoteService();
                    throw new RuntimeException(e);
                }).get();
        return properties.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey, entry -> {
                    if (entry.getValue() instanceof String str) {
                        return List.of(str);
                    } else if (entry.getValue() instanceof List<?> list) {
                        return list.stream().map(Object::toString).toList();
                    } else {
                        throw new RuntimeException("Unexpected type of value: " + entry.getValue());
                    }
                }));
    }

    private static Socket socket;
    private static RemoteService remoteService;

    private static RemoteService getRemoteService() throws IOException {
        if (remoteService != null) {
            return remoteService;
        } else {
            return createRemoteService();
        }
    }

    private static RemoteService createRemoteService() throws IOException {
        socket = new Socket("127.0.0.1", 6489);
        Launcher<RemoteService> launcher = Launcher.createLauncher(new Object(), RemoteService.class, socket.getInputStream(), socket.getOutputStream(), StandardIOLauncher.POOL, Function.identity());
        launcher.startListening();
        remoteService = launcher.getRemoteProxy();
        return remoteService;
    }

    private static void invalidateRemoteService() {
        if (socket != null) {
            try {
                socket.close();
            } catch (IOException e) {
                logger.error("Failed to close socket: {}", socket, e);
            } finally {
                socket = null;
            }
        }
        remoteService = null;
    }

    private interface RemoteService {
        @JsonRequest
        CompletableFuture<Map<String, Object>> query(String validExpr, boolean extras);
    }

}

File: zenscript-language-server/src/main/kotlin/raylras/zen/lsp/bracket/RpcClient.kt
package raylras.zen.lsp.bracket

import org.eclipse.lsp4j.jsonrpc.Launcher
import org.eclipse.lsp4j.jsonrpc.services.JsonRequest
import raylras.zen.lsp.StandardIOLauncher.POOL
import java.net.Socket
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit
import java.util.function.Function

object RpcClient {
    fun getEntryPropertiesRemote(validExpr: String): Map<String, Any> {
        return getRemoteService()
            .query(validExpr, true)
            .exceptionally { e ->
                invalidateRemoteService()
                throw e
            }
            .get(1, TimeUnit.SECONDS)
    }

    private var socket: Socket? = null
    private var remoteService: RemoteService? = null

    private fun getRemoteService(): RemoteService {
        if (remoteService == null) {
            createRemoteService()
        }
        return remoteService!!
    }

    private fun createRemoteService() {
        socket = Socket("127.0.0.1", 6489)
        val launcher = Launcher.createLauncher(
            Any(),
            RemoteService::class.java,
            socket!!.getInputStream(),
            socket!!.getOutputStream(),
            POOL,
            Function.identity()
        )
        launcher.startListening()
        remoteService = launcher.remoteProxy
    }

    private fun invalidateRemoteService() {
        socket?.close()
        socket = null
        remoteService = null
    }

    private interface RemoteService {
        @JsonRequest
        fun query(validExpr: String?, extras: Boolean): CompletableFuture<Map<String, Any>>
    }
}
